Mathlib/RingTheory/KrullDimension.lean

```diff
@@ -0,0 +1,123 @@
+
+
+## Main definitions and results
+
+* `ringKrullDim`: the ring theoretic Krull dimension of a commutative ring.
+
+* `ringKrullDim_eq_topologicalKrullDim`: the ring theoretic Krull dimension of a commutative ring
+is equal to the topological Krull dimension of its prime spectrum.
+
+* `ringKrullDim_eq_zero_of_field`: the Krull dimension of a field is zero.
+-/
+
+/--
+The ring theoretic Krull dimension is the Krull dimension of its spectrum ordered by inclusion.
+-/
+noncomputable def ringKrullDim (R : Type*) [CommRing R] : WithBot (WithTop ℕ) :=
+  krullDim (PrimeSpectrum R)
+
+variable {R S : Type*} [CommRing R] [CommRing S]
+
+@[nontriviality]
+lemma ringKrullDim_eq_bot_of_subsingleton [Subsingleton R] :
+    ringKrullDim R = ⊥ :=
+  krullDim_eq_bot_of_isEmpty
+
+lemma ringKrullDim_nonneg_of_nontrivial [Nontrivial R] :
+    0 ≤ ringKrullDim R :=
+  krullDim_nonneg_of_nonempty
+
+theorem ringKrullDim_eq_topologicalKrullDim :
+    ringKrullDim R = topologicalKrullDim (PrimeSpectrum R) :=
+  Eq.symm $ krullDim_orderDual.symm.trans $ krullDim_eq_of_orderIso
+  (PrimeSpectrum.pointsEquivIrreducibleCloseds R).symm
+
+/-- If `f : R →+* S` is surjective, then `ringKrullDim S ≤ ringKrullDim R`. -/
+theorem ringKrullDim_le_of_surjective (f : R →+* S) (hf : Function.Surjective f) :
+    ringKrullDim S ≤ ringKrullDim R :=
+  krullDim_le_of_strictMono (PrimeSpectrum.comap f) (Monotone.strictMono_of_injective
+    (fun _ _ hab ↦ Ideal.comap_mono hab) (PrimeSpectrum.comap_injective_of_surjective f hf))
+
+/-- If `I` is an ideal of `R`, then `ringKrullDim (R ⧸ I) ≤ ringKrullDim R`. -/
+theorem ringKrullDim_quotient_le (I : Ideal R) :
+    ringKrullDim (R ⧸ I) ≤ ringKrullDim R :=
+  ringKrullDim_le_of_surjective _ Ideal.Quotient.mk_surjective
+
+/-- If `R` and `S` are isomorphic, then `ringKrullDim R = ringKrullDim S`. -/
+theorem ringKrullDim_eq_of_ringEquiv (e : R ≃+* S) :
+    ringKrullDim R = ringKrullDim S :=
+  le_antisymm (ringKrullDim_le_of_surjective e.symm e.symm.surjective)
+    (ringKrullDim_le_of_surjective e e.surjective)
+
+/-- An alias for `ringKrullDim_eq_of_ringEquiv`. -/
+theorem _root_.RingEquiv.ringKrullDim (e : R ≃+* S) :
+    ringKrullDim R = ringKrullDim S :=
+  le_antisymm (ringKrullDim_le_of_surjective e.symm e.symm.surjective)
+    (ringKrullDim_le_of_surjective e e.surjective)
+
+section DimensionZero
+
+theorem _root_.PrimeSpectrum.asIdeal.isMaximal_of_ringKrullDim_eq_zero
+    (I : PrimeSpectrum R) (hdim : ringKrullDim R = 0) :
+    I.asIdeal.IsMaximal := by
+  have h := mem_maximals_of_krullDim_eq_zero hdim I
+  simp only [maximals, Set.top_eq_univ, Set.mem_univ, true_implies, true_and] at h
+  constructor; constructor
+  · exact Ideal.IsPrime.ne_top'
+  · intro J hIJ; by_contra hJ
+    obtain ⟨M, hM1, hM2⟩ := Ideal.exists_le_maximal J hJ
+    have := h $ show _ ≤ ⟨M, hM1.isPrime⟩ by exact (le_of_lt $ lt_of_lt_of_le hIJ hM2 : _ ≤ _)
+    exact (lt_self_iff_false _).1 $ lt_of_lt_of_le (lt_of_le_of_lt (by exact this) hIJ) hM2
```

Comment:
I think the next lemma is good enough.

Reply:
So shall I delete `PrimeSpectrum.asIdeal.isMaximal_of_ringKrullDim_eq_zero`?

Reply:
Yes that's my suggestion. Same below.